AudioSystem/SequenceAudioConfig.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+namespace DUCK.AudioSystem
+{
+	[CreateAssetMenu(menuName = "Sequence Audio Config (Fancy™)", order = 220)] // Right after Audio Mixer
+	public class AudioConfigSequence : AbstractAudioConfig
+	{
+		public enum AudioClipIndexType
+		{
+			Random,
+			Constant,
+			Parameter
+		}
+
+		private AudioConfig[] compositeAudioConfigs;
+		public IEnumerable<AudioConfig> GetAudioConfigs()
+		{
+			if (compositeAudioConfigs == null)
+			{
+				compositeAudioConfigs = entries.Select(e => e.AudioConfig).ToArray();
+			}
+
+			return compositeAudioConfigs;
+		}
+
+		[Serializable]
+		public class SequenceAudioConfigEntry
+		{
+			[SerializeField]
+			private AudioConfig audioConfig;
+
+			[SerializeField]
+			private AudioClipIndexType clipMode;
+
+			[SerializeField]
+			private int clipIndexConstant;
+
+			[SerializeField]
+			private string clipIndexParameter;
+
+			public AudioConfig AudioConfig { get { return audioConfig; } }
+
+			public int GetClipIndex(IDictionary<string, int> parameters = null)
+			{
+				switch (clipMode)
+				{
+					case AudioClipIndexType.Constant:
+						return clipIndexConstant;
+
+					case AudioClipIndexType.Parameter:
+						if (parameters != null && parameters.ContainsKey(clipIndexParameter))
+						{
+							return parameters[clipIndexParameter];
+						}
+						else goto default;
+
+					default:
+						return AudioConfig.RANDOM_CLIP;
+				}
+			}
+		}
+
+		public float DelayBetweenClips { get { return delay; } }
+
+		[SerializeField]
+		private SequenceAudioConfigEntry[] entries;
+
+		[SerializeField]
+		private float delay;
+
+		private Dictionary<string, int> clipIndexParameters;
+
+		public void SetParameter(string key, int value)
+		{
+			if (string.IsNullOrEmpty(key)) return;
+
+			if (clipIndexParameters == null)
+			{
+				clipIndexParameters = new Dictionary<string, int>();
+			}
+
+			clipIndexParameters[key] = value;
+		}
+
+		public int GetClipIndex(AudioConfig audioConfig)
+		{
+			var entry = entries.FirstOrDefault(e => e.AudioConfig == audioConfig);
```

Comment:
Personally, I'd like to reduce the LinQ usages as much as possible.

It generates a great amount of GC overhead. You will be surprised that how expensive of `FirstOrDefault` is.

Reply:
Interesting, I would have expected that LinQ is going to be performing faster or at least on par with standard methods. Which bit of the query is responsible for the memory overhead?

Reply:
Nan. LinQ in Unity is a bloody evil bastard.

In this case, `e => e.AudioConfig == audioConfig` creating an anonymous function (for predicate) FOR EACH element in the collection until it finds the right one. Those functions then will be thrown into the GC ocean and, one day, when the storm comes, your game lagged and the player fucked.